feat(i18n): initial setup

pavanjoshi914 opened the pull request:
Solves #600 

#### Type of change

 New feature (non-breaking change which adds functionality)


#### Describe the changes you have made in this PR -

setup i18next react-i18next browser-language-detector
create locale switcher
styling for both light and dark mode
translate heading
enable JSON imports in config
language persistence throughout the component
language persistence between different sessions
browser language detections
format JSON translation file using built-in eslint JSON formatted  and run automated using lintstaged




#### Screenshots of the changes (If any) -
![image](https://user-images.githubusercontent.com/55848322/159120558-337d0e05-1217-4a5b-9713-c34a18455458.png)
![image](https://user-images.githubusercontent.com/55848322/159120577-1adf227b-f192-4a93-86da-26a66b91c6f5.png)
![image](https://user-images.githubusercontent.com/55848322/159120591-d48559dc-e1ec-4e57-ad49-4bf541845afa.png)
![image](https://user-images.githubusercontent.com/55848322/159120610-579231d9-7589-4a1a-a598-25589eebcab1.png)

## Updates
locale switcher shifted under settings
![image](https://user-images.githubusercontent.com/55848322/159205681-26040fa4-addf-4915-ba83-7d93c3a03470.png)
locale switcher on Intro shown only under development mode
store locale in extension's storage as well

#### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration



#### Checklist:

- [X] My code follows the style guidelines of this project and performed a self-review of my own code
- [X] New and existing tests pass locally with my changes

pavanjoshi914 commented:
@bumi  @escapedcat initial setup for i18n

escapedcat commented on package.json:
```diff
@@ -34,6 +34,7 @@
     "storybook": "start-storybook -p 6006",
     "build-storybook": "build-storybook",
     "prepare": "husky install"
+    
```
This empty line can be removed

pavanjoshi914 replied:
done

escapedcat commented on src/app/components/LocaleSwitcher/LocaleSwitcher.tsx:
```diff
@@ -0,0 +1,25 @@
+import { useState, ChangeEvent } from "react";
```
Separate these into `import ...` and `import type ...`

pavanjoshi914 replied:
done

escapedcat commented on src/i18n/locales/en.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Can these json-file be formatted?

pavanjoshi914 replied:
yes its possible eslint provides built in json formatter. done

escapedcat commented on src/app/components/Navbar/Navbar.tsx:
```diff
@@ -39,6 +40,7 @@ export default function Navbar({ title, subtitle, children }: Props) {
         </div>
       )}
       <div className="md:w-4/12 lg:w-3/12 flex justify-end items-center">
+        <LocaleSwitcher />
```
Not sure if we need this in the navbar or if it would be ok to have this i.e. under settings

pavanjoshi914 replied:
this will be a case of dropdown under dropdown! mostly locale switchers shall be handy so I thought navbar shall be the best place! 
keeping it in settings and then it might happen that user never knows that we support multiple languages

bitcoinuser replied:
I prefer on settings too.

escapedcat replied:
Sure, I understand. Just thinking about this regarding things being stuffed in the navbar and if this is really needed or not. If the user is already using the browser in a certain language and alby is using that as a default then chances are high the user doesn't need to adjust the language all the time. Just wanted to give this a thought.  
@bumi @MoritzKa what do you think?

bumi replied:
was thinking the same. I guess the user does not change the language often. I'd prefer to save the user's language in the settings. If we can default to the browser's language then this should be fine, isn't it?

pavanjoshi914 replied:
okay so moving locale switcher under settings

pavanjoshi914 replied:
I moved the locale switcher under settings! shall we preserve the locale along with other settings?

pavanjoshi914 replied:
the problem I am getting is I am not able to do ```saveSettings``` from the intro page. where locale switcher is present as well.

pavanjoshi914 replied:
as we are using a browser language detector! so this will handle the case for new users! 

for existing users locale is persisted by default which is changed from language switcher, so do we really need to store locale in local browser storage?

escapedcat replied:
I think we can go with not storing it for now and adjust later if needed. Shouldn't block this.

pavanjoshi914 commented:
if you have gone through the documentation! either we can use [nesting of keys](https://gist.github.com/pavanjoshi914/a65af473209541c3a5cfecbac0f6e70f#using-nesting-of-keys)  or [react i18next namespaces](https://gist.github.com/pavanjoshi914/a65af473209541c3a5cfecbac0f6e70f#alternate-for-nesting-of-keys)

currently, I  am using namespace. if we wish to use nesting of keys as a key management strategy we can do that as well

escapedcat reviewed (commented):
Really cool, thanks! Some minor comments.  
@bumi do we want use [Weblate](https://weblate.org/) so others can help with further translations?

escapedcat commented on package.json:
```diff
@@ -33,11 +33,10 @@
   },
   "lint-staged": {
-    "src/**/*.{js,jsx,ts,tsx}": [
+    "src/**/*.{js,jsx,ts,tsx,json}": [
```
🙌

escapedcat commented on src/app/components/LocaleSwitcher/LocaleSwitcher.tsx:
```diff
@@ -0,0 +1,26 @@
+import { useState } from "react";
+import type { ChangeEvent } from "react";
+import i18n from "i18next";
+// import i18n (needs to be bundled ;))
```
I don't understand this comment

pavanjoshi914 replied:
we use a single JSON file per locale, but when react i18next bundles everything into a single object {en: {} , hi:{}} something like that. so that comment is pointing something like this.

escapedcat replied:
Ah ok, do we need that comment though? Does it explain anything the import doesn't?

pavanjoshi914 replied:
not much but standard documentation keeps this so didn't change that. if you wish I can remove that for sure

escapedcat replied:
Ah I see, it's from here:  
https://react.i18next.com/latest/using-with-hooks  
I think it's good to have it in the docs, but it doesn't add any value here.

pavanjoshi914 replied:
removed that

escapedcat reviewed (approved):
Awesome job @pavanjoshi914 !

pavanjoshi914 commented:
Squashed all the commits into a single commit. @bumi  PR is ready to be merged!

pavanjoshi914 commented:
> Really cool, thanks! Some minor comments. @bumi do we want to use [Weblate](https://weblate.org/) so others can help with further translations?

regarding weblate! currently, we can just focus on completing i18n with a single alternate language. setup for continuous localization can be done once we are finished with this
https://gist.github.com/pavanjoshi914/a65af473209541c3a5cfecbac0f6e70f#continuous-localization

escapedcat commented:
Applied for the "Libre hosting" plan on weblate already

bumi reviewed (changes requested):
I think we should save the selected language in the settings storage. Otherwise it will get lost after the app reloads. 

You can see how to save the settings there: https://github.com/getAlby/lightning-browser-extension/blob/4a8d6336acdbdd2a0848426d404cf79a71f7404b/src/app/screens/Settings.tsx#L42

pavanjoshi914 commented:
> I think we should save the selected language in the settings storage. Otherwise it will get lost after the app reloads.
> 
> You can see how to save the settings there:
> 
> https://github.com/getAlby/lightning-browser-extension/blob/4a8d6336acdbdd2a0848426d404cf79a71f7404b/src/app/screens/Settings.tsx#L42

sure so we will have incoming locale from three places
1. browser
2. language switcher
3. settings

Case 1. storing browser detected language or fallback language when language is not available 
when a user installs the app we let the browser detector, detect the language and use that language. either browser detected language will be used and if not available fallback language ```:en ``` will be used

```resolvedLanguage``` can handle this case, we exported the final resolved language under default settings

```
export const DEFAULT_SETTINGS = {
  websiteEnhancements: true,
  userName: "",
  locale: i18n.resolvedLanguage,
};
```

Case 2:  setting language changed via locale switcher

1. locale switcher under settings
passing saveSetting function as props and executed in ```LocaleSwitcher.tsx```

```
<LocaleSwitcher saveSettings={saveSetting} />
```
```
 if (dropdownLang !== newLanguage) {
      setDropdownLang(newLanguage);
      i18n.changeLanguage(newLanguage);
      props.saveSettings({ locale: newLanguage });
    }
```

2. locale switcher under intro section

I can't pass a function as props for this locale switcher. one idea I think is refactor and make function accessible


Before I knew just the basics of react before, I am still learning :)  what can be the best possible approach for handling case 2 @bumi  @escapedcat  any ideas?

bumi commented:
Can you elaborate on `2.`? Why doesn't it work the same way as in the settings on the welcome page?

pavanjoshi914 commented:
> Can you elaborate on `2.`? Why doesn't it work the same way as in the settings on the welcome page?

I made a  silly mistake! I thought ```saveSetting``` is the function which is providing save functionality and I was trying this to pass as props in localeSwitcher component. it was just a wrapper for ```const response = await api.setSetting(setting);	``` 
```
async function saveSetting(
    setting: Record<string, string | number | boolean>
  ) {
    const response = await api.setSetting(setting);
    setSettings(response);
  }
  ```
  
  did required changes @bumi @escapedcat kindly review

escapedcat commented:
> I made a silly mistake

![image](https://user-images.githubusercontent.com/1016218/159872764-36466783-da4a-4b4e-9e41-b92a4d14a20f.png)

bumi commented on src/app/router/Welcome/Welcome.tsx:
```diff
@@ -73,10 +76,10 @@ function App() {
       <div className="relative max-w-7xl mx-auto px-4 sm:px-6 lg:px-8">
         <div className="text-center font-serif font-medium text-2xl pt-7 pb-3 dark:text-white">
-          <p>The power of lightning in your browser</p>
+          <p>{t("heading")}</p>
```
what is the convention for naming these keys? 
i18n automatically nests these keys under the component names? (do we need to document this for translators?)

escapedcat replied:
I think it's a combination with `useTranslation(["welcome"])` in line 53.  
This is a short way of writing `welcome.heading`.

bumi reviewed (commented):
From the code I don't fully understand where we load the language from the settings. 
Can you point me to that? 

Also how do we roll this out? Should we only show the language switcher in dev mode until we have a decent amount of the app translated?

pavanjoshi914 commented:
> if you have gone through the documentation! either we can use [nesting of keys](https://gist.github.com/pavanjoshi914/a65af473209541c3a5cfecbac0f6e70f#using-nesting-of-keys) or [react i18next namespaces](https://gist.github.com/pavanjoshi914/a65af473209541c3a5cfecbac0f6e70f#alternate-for-nesting-of-keys)
> 
> currently, I am using namespace. if we wish to use nesting of keys as a key management strategy we can do that as well

...

bumi commented:
sorry, my bad. forgot...sure. really good, let's add that to the wiki? and add a link in the i18config in code to that wiki page?

bumi commented:
(ups and I don't even hit the right button this morning, sorry. :) )

pavanjoshi914 commented:
> sorry, my bad. forgot...sure. really good, let's add that to the wiki? and add a link in the i18config in code to that wiki page?

sure , but before that i have to do few revisions to make it more formal.

escapedcat commented:
> Also how do we roll this out? Should we only show the language switcher in dev mode until we have a decent amount of the app translated?

Good point. We can also add `alpha` to the settings-label, already show it and open an issue to move all texts step by step to translation-strings.

pavanjoshi914 commented:
> From the code I don't fully understand where we load the language from the settings. Can you point me to that?
> 
> Also how do we roll this out? Should we only show the language switcher in dev mode until we have a decent amount of the app translated?


> > I made a silly mistake
> 
> ![image](https://user-images.githubusercontent.com/1016218/159872764-36466783-da4a-4b4e-9e41-b92a4d14a20f.png)
haha nice one @escapedcat 😁

AaronDewes commented:
@bumi If you're looking for a translation platform, I recommend hosted.weblate.org, they are free for open source projects if you apply. I'm using it for Citadel and it works great

pavanjoshi914 commented:
so any special use case do you have in mind @bumi? which supports storing locale in the custom objects along with other settings and loading locale from settings?
if not then we shall not interrupt the normal flow of the plugin and do something as shown below! WDYT

```
 lng: localStorage.getItem(I18N_LANGUAGE) || "en"
```

pavanjoshi914 commented:
> @bumi If you're looking for a translation platform, I recommend hosted.weblate.org, they are free for open source projects if you apply. I'm using it for Citadel and it works great

thanks for the suggestion @AaronDewes ! yes we are already planned to use weblate for continuous localization

bumi commented:
@pavanjoshi914 hmm, I mean we now save the settings, so we should also use that. 
Also I am not sure how good it is to use the localstorage in the extension. for that extensions have some extension storage (which we use to save the settings)

@AaronDewes nice, thanks. good to hear you're using and recommending it.

pavanjoshi914 commented:
> 

sure doing that then! @escapedcat  WDYT

escapedcat commented:
I think if there are no issues we should go with what the plugin does. At least for now.  
Would be great to get this in and start translating.

bumi commented:
I think I am missing something...
IF we use the default locale detector WHY do we then have the setting? that setting now has no effect.

And I think in the extension context it is recommended to use the [extension storage](https://developer.chrome.com/docs/extensions/reference/storage/) I would also prefer not to store things in different places.

bumi commented:
As discussed in the call: 
* we leave it as is and make a follow up PR that loads the locale from the our settings ([custom detector example code](https://github.com/i18next/i18next-browser-languageDetector/issues/140)) (@pavanjoshi914 can you create an issue for that?
* we show the language switcher on the welcome/intro screens ONLY in dev mode (or remove it for now) - because we do not have any translations, yet.
* we add something like "work in progress - help us translate Alby in your language" note to the locale switcher setting

pavanjoshi914 commented:
@bumi  @escapedcat

pavanjoshi914 commented:
> As discussed in the call:
> 
> * we leave it as is and make a follow up PR that loads the locale from the our settings ([custom detector example code](https://github.com/i18next/i18next-browser-languageDetector/issues/140)) (@pavanjoshi914 can you create an issue for that?
> * we show the language switcher on the welcome/intro screens ONLY in dev mode (or remove it for now) - because we do not have any translations, yet.
> * we add something like "work in progress - help us translate Alby in your language" note to the locale switcher setting



> As discussed in the call:
> 
> * we leave it as is and make a follow up PR that loads the locale from the our settings ([custom detector example code](https://github.com/i18next/i18next-browser-languageDetector/issues/140)) (@pavanjoshi914 can you create an issue for that?
> * we show the language switcher on the welcome/intro screens ONLY in dev mode (or remove it for now) - because we do not have any translations, yet.
> * we add something like "work in progress - help us translate Alby in your language" note to the locale switcher setting

https://github.com/getAlby/lightning-browser-extension/issues/743

bumi commented:
@dylancom can you have a quick look and merge it when approved?

bumi reviewed (approved):
utACK

pavanjoshi914 commented:
any updates here?

dylancom commented:
@pavanjoshi914 Going to test it now!

dylancom commented:
The vertical alignment of the LanguageSwitcher seems off.
We already have a <Select /> component by the way that you can use from components/form
<img width="149" alt="Schermafbeelding 2022-03-29 om 16 22 19" src="https://user-images.githubusercontent.com/12894112/160633549-7132219d-c111-44d2-be1d-46bc34dc5e1f.png">

pavanjoshi914 commented on src/app/components/LocaleSwitcher/LocaleSwitcher.tsx:
```diff
@@ -0,0 +1,27 @@
+import api from "../../../common/lib/api";
+
+export default function LocaleSwitcher() {
+  const [dropdownLang, setDropdownLang] = useState(i18n.language || "en");
```
done

pavanjoshi914 commented on src/app/components/LocaleSwitcher/LocaleSwitcher.tsx:
```diff
@@ -0,0 +1,27 @@
+import { useState } from "react";
+import type { ChangeEvent } from "react";
+import i18n from "i18next";
+import "../../../i18n/i18nConfig";
```
done

pavanjoshi914 commented on src/app/router/Welcome/Welcome.tsx:
```diff
@@ -72,11 +75,15 @@ function App() {
-      {process.env.NODE_ENV === "development" && <DevMenu />}
-
+      {process.env.NODE_ENV === "development" && (
+        <Fragment>
```
done

pavanjoshi914 commented:
@dylancom

escapedcat commented:
> > The vertical alignment of the LanguageSwitcher seems off. It also doesn't show the pointer cursor while hovering over it. We already have a component by the way that you can use from components/form
> 
> not sure what I have to do here! any screenshot which can help to understand where is the problem?

I think he meant you can use this component instead of a normal select:  
https://github.com/getAlby/lightning-browser-extension/blob/test/select-account-after-account-delete/src/app/components/form/Select/index.tsx  

You can find an example how to use it here:
- https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/components/form/Select/Select.stories.tsx
- https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/screens/Receive.tsx/#L199-L214
  Funnily it looks like this is commented out so we do not actually use this component anywhere atm. But still it would be good to use it to make sure future selects are aligned

pavanjoshi914 commented:
> > > The vertical alignment of the LanguageSwitcher seems off. It also doesn't show the pointer cursor while hovering over it. We already have a component by the way that you can use from components/form
> > 
> > 
> > not sure what I have to do here! any screenshot which can help to understand where is the problem?
> 
> I think he meant you can use this component instead of a normal select: [https://github.com/getAlby/lightning-browser-extension/blob/test/select-account-after-account-delete/src/app/components/form/Select/index.tsx](https://github.com/getAlby/lightning-browser-extension/blob/test/select-account-after-account-delete/src/app/components/form/Select/index.tsx?rgh-link-date=2022-04-01T08%3A15%3A21Z)
> 
> You can find an example how to use it here:
> 
> * [https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/components/form/Select/Select.stories.tsx](https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/components/form/Select/Select.stories.tsx?rgh-link-date=2022-04-01T08%3A15%3A21Z)
> * [https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/screens/Receive.tsx/#L199-L214](https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/screens/Receive.tsx/?rgh-link-date=2022-04-01T08%3A15%3A21Z#L199-L214)
>   Funnily it looks like this is commented out so we do not actually use this component anywhere atm. But still it would be good to use it to make sure future selects are aligned

already using! but i refactored by default class present with component!

pavanjoshi914 commented:
> already using! but i refactored by default class present with component!

the only important class was ```h-14``` in that default class (```w-full h-14 border-gray-300 rounded-md shadow-sm text-gray-700 bg-white```), added that too! not sure if this is what Dylan willing to have

escapedcat commented:
> already using! but i refactored by default class present with component!

👍  Ah sorry, didn't see.

Might be good having default styles on a select which can be extended or overwritten, but we can take care of that later on.

pavanjoshi914 commented:
> > already using! but i refactored by default class present with component!
> 
> 👍 Ah sorry, didn't see.
> 
> Might be good having default styles on a select which can be extended or overwritten, but we can take care of that later on.

yup but the class has ```w-full``` which we don't want in case of language switcher also tried overwriting it! nothing happened

dylancom commented:
@pavanjoshi914 To give the <Select /> a certain width, you can just wrap it in a div with a specified width...
Our form inputs also work like that.

pavanjoshi914 commented:
> 

w-full covers entire screen
![image](https://user-images.githubusercontent.com/55848322/161264679-3a17335c-1dbb-4e85-a75d-8340071b17c3.png)

pavanjoshi914 commented:
> @pavanjoshi914 What is the problem with the w-full class? You let the width of the component to be decided by the parent that it's wrapped in. It might need some updated styling if it's too high now, as it has been styled a while ago and not used in prod yet. But I would keep default styles instead of passing in custom classes, and let a parent element decide how it's aligned, what it's width is etc.

so I can wrap the select element in div tag to overwrite few of the styles.

dylancom commented:
w-full means width: 100%; which means take 100% width of the parent container. So if your parent container is 400px wide, w-full means a width of 400px...

dylancom commented:
If your screen width is smaller than 1300px in the Welcome screen, then you can't click the language switcher. (elements overlapping)

pavanjoshi914 commented:
@bumi kindly review and merge
